llvm/lib/SYCLLowerIR/LowerWGScope.cpp

```diff
@@ -752,6 +780,10 @@ PreservedAnalyses SYCLLowerWGScopePass::run(Function &F,
                                             FunctionAnalysisManager &FAM) {
   if (!F.getMetadata(WG_SCOPE_MD))
     return PreservedAnalyses::all();
+  // If a function does not have any PFWI calls and it has calls to a function
+  // that has work_group metadata, then we do not need to lower such functions.
+  if (!hasPFWICall(F) && hasCallToAFuncWithWGMetadata(F))
```

Comment:
can you explain why we can bail early in this case? thanks.

Reply:
This pass tries to lower parallel_for_work_group and parallel_for_work_item markings. Any code that occurs between the two markings need to be handled in a special way which is what this pass mainly does.
There are multiple possibilities for functions processed here:

CASE 1
void foo(...) { // No work group scope marking
}
--> We exit early based on Line 781 check

CASE 2
// bar(....) { // Work group scope marking
// CODE TO BE PROCESSED
// work item
}
void foo(...) { // Work group scope marking
....
// no work items
// call to bar

}
--> In this case we want to only process the code inside bar. We do not need to process code inside foo as it just transmits all values to bar.

CASE 3
void foo(...) { // work group scope marking
// No work item
}
--> We process foo here even though we do not have any work items. This is sometimes that we have to support as the documentation is unclear about this. Existing tests rely on this support.

Sorry if the explanation is a bit vague. I tried to explain it in the best way without getting too verbose.

Thanks

Reply:
got it, thanks